### ISSUE TITLE
Languages whose syntax includes "//" cannot be correctly defined

### ISSUE THREAD

**Issue by dlucanu** [kframework/k-legacy#694]
The following definition includes "//" as a syntax construct:

```
module TEST
   syntax Cmd ::= Id | "//" Id
   syntax Cmds ::= List{Cmd, ""}
endmodule
```

If we compile it and execute the program "abc // def" then we get the following output:

```
<k>
    abc
</k>
```

We see that the part "// def" disappeared because it was treated as a comment.
The question is: how to define the language construct "//"? 
The language designer strongly wishes to have this construct and a different syntax for comments.

**Comment by dwightguth** [comment]
This is a good feature to have, we just need someone with the bandwidth to
work on it.

I believe the correct solution is to have no comments by default in the
program parser, and a special nonterminal which gets added to the layout
allowed between items in a production. So for example:

syntax COMMENT ::= Token{"//.*\n"}

**Comment by radumereuta** [comment]
There is no easy solution for this, and I don't have a plan to do it any time soon.
The whitespace problem is a tricky one in SDF.
Hopefully I can do something about that in the new parser. I have some ideas about that, but it's going to be one of the last things to do.

**Comment by dlucanu** [comment]
This is very bad because now i am not able to supply a solution to guys from INRIA Lille.
They want their programs to be executed with K, compiled in Forth and other machine language. 
We must give to the user the flexibility to define any language she wants.

**Comment by grosu** [comment]
We agree, but the question is who wants to work on this.  Everything we do right now is under "emergency" status.  It looks like it falls into Radu's teritory.  However, since we intend to change the SDF parser very soon, then it seems to be little motivation to fix this problem with SDF.  On the other hand, the current parser has much bigger problems and priorities.  Unfortunately, we only have Radu work on the front end these days.  Anybody volunteers to help with the front end?

**Comment by radumereuta** [comment]
The only solution I can give right now, is an external parser.
Also, having someone else familiar with the front end could be useful in the future.

**Comment by traiansf** [comment]
A simple preprocessor only replacing // with something like "/ /" or "||"
and only using /\* */ for comments would probably be a reasonable solution.

2014-07-02 21:38 GMT+03:00 Radu Mereuta notifications@github.com:

**Comment by dlucanu** [comment]
I agree to use "/_..._/" for comments but I have to tell somehow to the compiler to disable the processing of "//" as comments. This should be doable.

**Comment by traiansf** [comment]
I was saying to use a preprocessor before sending the programs to krun

2014-07-02 22:08 GMT+03:00 Dorel Lucanu notifications@github.com:

**Comment by dlucanu** [comment]
That means to give semantics to "||" instead of "//" in the K definition, which is a bit confusing.
It could be  a solution only for a short term (I already applied it ...)

**Comment by dwightguth** [comment]
This is very very easy to implement in the new parser, so I am going to assign it to Radu. Radu, the solution here is to create a production in kast.k which is the following:

```
syntax #Layout ::= r"((/\\*([^\\*]|(\\*+([^\\*/])))*\\*+/)|(//[^\n\r]*)|([\\ \n\r\t]))*" [token]
```

And then instead of adding a RegExState to the grammar followed by a  DeleteState, we add a NonTerminalState followed by a DeleteState. then we can probably add something to the compiler specifying that if it cannot find a declaration anywhere in the module for the #Layout sort, the module containing the default layout is implicitly imported into the current parser.

**Comment by radumereuta** [comment]
Well... we already had a discussion about this problem. See the paragraph at the end: https://github.com/kframework/k/wiki/Syntax-specification-for-the-new-parser

**Comment by dwightguth** [comment]
This is essentially the same proposal except my solution is conceptually
cleaner and doesn't require whitespace to be special-cased in Grammar.java.
It is also more in line with the mechanism already used to specify default
configuration.
